### PR TITLE
Issue #2656896 Product revisions

### PR DESCRIPTION
**Pull request by czigor**
https://www.drupal.org/node/2656896
First stab, untested. Not worth looking at it yet.

### CODE REVIEW

File: modules/product/src/Entity/Product.php [hunk 1]
```diff
@@ -285,6 +290,43 @@ protected function ensureTranslations(array $entities) {
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function preSaveRevision(EntityStorageInterface $storage, \stdClass $record) {
+    parent::preSaveRevision($storage, $record);
+
+    if (!$this->isNewRevision() && isset($this->original) && (!isset($record->revision_log) || $record->revision_log === '')) {
+      // If we are updating an existing node without adding a new revision, we
```
Comment (borisson) [inline modules/product/src/Entity/Product.php]: s/node/product/

File: modules/product/src/Entity/ProductType.php [hunk 2]
```diff
@@ -74,6 +75,13 @@ class ProductType extends ConfigEntityBundleBase implements ProductTypeInterface
   protected $description;
 
   /**
+   * Default value of the 'Create new revision' checkbox of this node type.
```
Comment (borisson) [inline modules/product/src/Entity/ProductType.php]: s/node/product/

File: modules/product/src/Form/ProductForm.php [hunk 3]
```diff
@@ -84,6 +97,41 @@ public function form(array $form, FormStateInterface $form_state) {
...
+      '#type' => 'details',
+      '#group' => 'advanced',
+      '#title' => t('Revision information'),
+      // Open by default when "Create new revision" is checked.
+      '#open' => $product->isNewRevision(),
+      '#attributes' => array(
+        'class' => array('node-form-revision-information'),
+      ),
+      '#attached' => array(
+        'library' => array('node/drupal.node'),
```
Comment (borisson) [inline modules/product/src/Form/ProductForm.php]: If we really need this, we should depend on node from the product module. As far as I can tell, we only need one part of the javascript related to this library, so we should probably copy only that part over into a new library?

File: modules/product/src/Form/ProductForm.php [hunk 4]
```diff
@@ -84,6 +97,41 @@ public function form(array $form, FormStateInterface $form_state) {
...
+      ),
+      '#weight' => 20,
+      '#optional' => TRUE,
+    );
+
+    $form['revision'] = array(
+      '#type' => 'checkbox',
+      '#title' => t('Create new revision'),
+      '#default_value' => $product->type->entity->isNewRevision(),
+      '#access' => $current_user->hasPermission('administer nodes'),
```
Comment (borisson) [inline modules/product/src/Form/ProductForm.php]: s/nodes/products/

File: modules/product/src/Form/ProductForm.php [hunk 5]
```diff
@@ -270,6 +318,32 @@ protected function actions(array $form, FormStateInterface $form_state) {
 
   /**
    * {@inheritdoc}
+   *
+   * Updates the node object by processing the submitted values.
+   *
+   * This function can be called by a "Next" button of a wizard to update the
+   * form state's entity with the current step's values before proceeding to the
+   * next step.
+   */
```
Comment (borisson) [inline modules/product/src/Form/ProductForm.php:320]: It's not allowed to have both `{@inheritdoc}` and documentation in docblock, so we should either remove this comment or copy the relevant portion of the parent's docblock over.

File: modules/product/src/Form/ProductForm.php [hunk 6]
```diff
@@ -270,6 +318,32 @@ protected function actions(array $form, FormStateInterface $form_state) {
...
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    // Build the node object from the submitted values.
```
Comment (borisson) [inline modules/product/src/Form/ProductForm.php]: s/node/product/

**Comment by czigor** [comment]
Ready for review now.

File: modules/product/src/Entity/ProductType.php [hunk 7]
```diff
@@ -90,6 +98,20 @@ class ProductType extends ConfigEntityBundleBase implements ProductTypeInterface
...
    * {@inheritdoc}
    */
+  public function isNewRevision() {
+    return $this->new_revision;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setNewRevision($new_revision) {
```
Comment (andypost) [inline modules/product/src/Entity/ProductType.php:108]: no external interfaces changed so where this method from?

**Comment by bojanz** [comment]
The issue (http://drupal.org/node/2656896) has a newer patch, let's continue there.

**Pull request closed without merging**